simln-lib/src/lib.rs

```diff
@@ -1031,257 +1056,194 @@ impl<C: Clock + 'static> Simulation<C> {
-                let node_info = ce_node.lock().await.get_info().clone();
-                log::debug!("Starting events consumer for {}.", node_info);
-                if let Err(e) =
-                    consume_events(ce_node, clock, receiver, ce_output_sender, ce_listener).await
-                {
-                    ce_shutdown.trigger();
-                    log::error!("Event consumer for node {node_info} exited with error: {e:?}.");
-                } else {
-                    log::debug!("Event consumer for node {node_info} completed successfully.");
-                }
-            });
-        }
-
-        channels
-    }
-
     /// Responsible for spinning up producers for a set of activities. Requires that a consumer channel is present
     /// for every source node in the set of executors.
     async fn dispatch_producers(
         &self,
-        executors: Vec<ExecutorKit>,
-        producer_channels: HashMap<PublicKey, Sender<SimulationEvent>>,
+        mut executors: Vec<ExecutorKit>,
+        output_sender: Sender<SimulationOutput>,
         tasks: &TaskTracker,
     ) -> Result<(), SimulationError> {
-        for executor in executors {
-            let sender = producer_channels.get(&executor.source_info.pubkey).ok_or(
-                SimulationError::RandomActivityError(RandomActivityError::ValueError(format!(
-                    "Activity producer for: {} not found.",
-                    executor.source_info.pubkey,
-                ))),
-            )?;
-
-            // pe: produce events
-            let pe_shutdown = self.shutdown_trigger.clone();
-            let pe_listener = self.shutdown_listener.clone();
-            let pe_sender = sender.clone();
-            let clock = self.clock.clone();
-
-            tasks.spawn(async move {
-                let source = executor.source_info.clone();
+        let mut heap: BinaryHeap<Reverse<PaymentEvent>> = BinaryHeap::new();
+        let payment_event_payloads: Arc<Mutex<HashMap<PublicKey, PaymentEventPayload>>> =
+            Arc::new(Mutex::new(HashMap::new()));
 
-                log::info!(
-                    "Starting activity producer for {}: {}.",
-                    source,
-                    executor.payment_generator
-                );
-
-                if let Err(e) = produce_events(
-                    executor.source_info,
-                    executor.network_generator,
-                    executor.payment_generator,
-                    clock,
-                    pe_sender,
-                    pe_listener,
-                )
+        // in order to choose a deterministic destination is necessary to sort the nodes by its public key
+        executors.sort_by(|ex1, ex2| ex1.source_info.pubkey.cmp(&ex2.source_info.pubkey));
+        for executor in executors {
+            let payload = PaymentEventPayload {
+                wait_time: Duration::from_secs(0),
+                destination: executor.source_info.clone(),
```

Comment:
Setting a field called destination to the source value and then overwriting it later is some serious code smell to me.